aten/src/ATen/core/ivalue_inl.h

```diff
@@ -522,6 +527,30 @@ c10::optional<T> generic_to(
   return std::move(ivalue).to<T>();
 }
 
+namespace detail {
+template <typename Tuple, std::size_t... I>
+Tuple generic_to_tuple_impl(
+    const std::vector<IValue>& t,
+    c10::guts::index_sequence<I...>) {
+  return std::make_tuple(
+      t[I].to<typename std::tuple_element<I, Tuple>::type>()...);
+}
+}
+
+template <
+    typename... Args,
+    typename Indices = c10::guts::make_index_sequence<sizeof...(Args)>,
+    c10::guts::enable_if_t<
+        !c10::guts::disjunction<
+            std::is_lvalue_reference<Args>...,
+            c10::guts::negation<std::is_constructible<IValue, Args>>...>::value,
+        std::nullptr_t> = nullptr>
+std::tuple<Args...> generic_to(IValue ivalue, _fake_type<std::tuple<Args...>>) {
+  auto vals = ivalue.toTuple()->elements();
+  TORCH_INTERNAL_ASSERT(vals.size() == sizeof...(Args));
```

Comment:
Hmm... an assert? The existing guidance in this file is a bit murky: it is true most of the type tests are done using `AT_ASSERT`, but that old mean actually means that we haven't appropriately applied the guidance from https://github.com/pytorch/pytorch/issues/20287

I think ivalue tests should be done with `TORCH_CHECK`. The reasoning is that IValue is part of the public C++ API, and a user may very well attempt to extract an IValue at the wrong type. IT would be wrong for them to report an error message to PyTorch in that case.